add calendar and fetch recent meetings in a month

subhangi2731 opened the pull request:
Added the calendar `fullcalendar.io` plugin and fetched the data of all the recent meetings in the month grid view. When the calendar page is loaded the function is called and rendered to fetch recent meetings. However, will display the recent meetings in the calendar also as events.

closes #228

subhangi2731 commented:
![Screenshot from 2021-12-23 21-19-25](https://user-images.githubusercontent.com/53938490/147264607-3e147fcf-b014-46e8-82da-5351bf3990f8.png)
![Screenshot from 2021-12-23 21-19-16](https://user-images.githubusercontent.com/53938490/147264617-22ccbcb2-fd5f-4125-b376-072949949aea.png)

subhangi2731 commented:
@darknao @t0xic0der Please review and let me know. I have displayed the meetings of this current month and the meeting title redirects to the meeting details page. For the previous month, I will display the meetings by tomorrow. Also please let me know how the UI looks

t0xic0der reviewed (changes requested):
Please add `fedora_messaging` to the `requirements.txt`

t0xic0der reviewed (changes requested):
Please widen the navbar controls to occupy the width taken up by the container that follows below.

For instance, the following 

![image](https://user-images.githubusercontent.com/49605954/147311755-ecca679d-aa6c-458d-a448-0cd1fa375914.png)

Should look like the following

![image](https://user-images.githubusercontent.com/49605954/147311795-ce8c3fe7-5963-406a-a592-d135f72dfc29.png)

t0xic0der reviewed (changes requested):
Please remove or reduce the margin at the top of the container.

For instance, the following

![image](https://user-images.githubusercontent.com/49605954/147312028-a71f6327-c920-4c80-8767-e1686d9f9d74.png)

Should look, more or less, like the following 

![image](https://user-images.githubusercontent.com/49605954/147312064-666b96e2-4b66-472c-8f82-3e69ec15e3e6.png)

Prefer the use of the inbuilt margin classes like `mt-$(size)` over styles like `margin-top: $(size)%` as the latter change when the width of the viewport is varied while the former doesn't.

t0xic0der reviewed (changes requested):
Please remove the arrow buttons on the top right corner of the calendar as the recent view is only supposed to show the meetings which took place in this month. Also, consider adding a week view like we discussed before to better filter the meetings.

@mairin, one question. Should we keep the month view as a default or the week view when it comes to the recent meetings page?

![image](https://user-images.githubusercontent.com/49605954/147312293-e8b8ea16-3c06-4d6a-a897-f89fcd8e18b0.png)

t0xic0der reviewed (changes requested):
There's a significant amount of delay from the time when the page is rendered to the time when the calendar along with the list of meetings appear.

So when I load up the page, I get to see the following for a couple of seconds.

![Screenshot from 2021-12-24 09-04-02](https://user-images.githubusercontent.com/49605954/147313011-43f1e47d-c7f5-49ba-9ac0-fffe4237a28f.png)

Before the following comes

![Screenshot from 2021-12-24 08-57-24](https://user-images.githubusercontent.com/49605954/147313034-162a1935-9e19-4cf7-bee5-594238d28bda.png)

Now, two-three seconds is absolutely not the end of the world but I am on a fast internet connection and assuming the similar conditions for the entire community using this application would be very wrong on my part.

Hence I would ask you to do two things.

1. Render the empty calendar first on the page, before even reaching out to Datagrepper to get a list of recent meetings. 
2. Show an indeterminate spinner on the page somewhere to let the users know that the list of recent meetings are being fetched and the calendar would be soon populated with those.
   Take for instance, https://github.com/fedora-infra/mote/blob/main/src/mote/static/jscn/fragment.js#L103-L105 of the `populate_channel_list()` function has an indeterminate spinner rendered by default when the modal is loaded up and it is soon replaced/removed when the contents have arrived.

t0xic0der reviewed (changes requested):
Please add another link on the navbar and keep it as active to mark that clicking on it would lead to the current page.

For example, the link may be called `Recent meetings` and it can be marked as `active` (by using the said class). It can act as both a title for the current page and an indicator that the landing page would lead to the page which shows recent meetings.

Here's an indicative example.

![image](https://user-images.githubusercontent.com/49605954/147313368-cc4d4515-2f78-4611-9417-c53e936d11c5.png)

t0xic0der reviewed (changes requested):
Please stick with the design that @mairin suggested here https://github.com/fedora-infra/mote/issues/228#issuecomment-985007664. I am attaching the same image here for a quick reference.

![image](https://user-images.githubusercontent.com/49605954/147313504-5f1ef46d-fad0-4b96-a04b-81d015b29f39.png)

fullcalendar.io is very extensible and customizable when it comes to styling to I do not think it would be difficult to closely adhere to the design suggestions that she made.

t0xic0der reviewed (changes requested):
Hmm. The `today` button is inactive. Any plans for the same?

![image](https://user-images.githubusercontent.com/49605954/147313737-e2a9c94f-bb0e-4448-a0fb-fe264b2dfd5e.png)

subhangi2731 commented:
@t0xic0der I have done the changes as said. I was thinking can we keep the arrow buttons to navigate between the weeks of the current month and when we are on week view and go to previous weeks the `today` button gets activated which on clicking quickly helps navigate to the current date. Please let me know your opinions also the design suggested by @Mairin contains Fedora Bootstrap implementation for that tab icons and for these npm installation is required. Let me know your suggestions

subhangi2731 commented:
![Screenshot from 2021-12-25 13-20-45](https://user-images.githubusercontent.com/53938490/147380307-82c8249a-8ac0-41a5-8686-d10b10bb5749.png)
![Screenshot from 2021-12-25 13-20-31](https://user-images.githubusercontent.com/53938490/147380308-c549335e-067c-4154-9783-092253ab5ecc.png)

t0xic0der reviewed (changes requested):
The suggestion of the links on the navbar to occupy 100% of the width seems to have been misinterpreted.

It looks like this now.

![image](https://user-images.githubusercontent.com/49605954/147430931-6cfe4c53-c240-4a46-92bb-08c44801820f.png)

Please take a "closer" look at my review and the screenshots I have attached at https://github.com/fedora-infra/mote/pull/232#pullrequestreview-839714652. The code in the expected form screenshot give enough information on what needs to be done.

t0xic0der reviewed (changes requested):
> Please let me know your opinions also the design suggested by @mairin contains Fedora Bootstrap implementation for that tab icons and for these npm installation is required. Let me know your suggestions


There's no such need to do anything of that sort. Please make use of custom CSS classes to make the frontend resemble with the design that she proposed. Fedora Bootstrap is based on Bootstrap 4.3.1 and we would not want to make use of it as Bootstrap 5 (the one that we currently use) is not backwards-compatible with Bootstrap 4.x.

t0xic0der commented on src/mote/templates/mainpage.html:
```diff
@@ -33,11 +33,13 @@
         <script src="{{ url_for('static', filename='jscn/fragment.js') }}"></script>
         <script src="{{ url_for('static', filename='jscn/easy.qrcode.min.js') }}"></script>
         <script src="{{ url_for('static', filename='jscn/socket.io.js') }}"></script>
+        <link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/fullcalendar@5.10.1/main.css">
+    <script src="https://cdn.jsdelivr.net/npm/fullcalendar@5.10.1/main.js"></script>
```
Please download them into the `/static/` subdirectory and load them up from there.

Also, rename `main.css` and `main.js` aptly to be representative of what they do - say, call them `fullcalendar.css` and `fullcalendar.js` (or something like that).

t0xic0der commented on src/mote/templates/mainpage.html:
```diff
@@ -72,12 +79,12 @@
         </nav>
         <div class="container-lg">
-            <div class="card text-center" style="margin-top: 10%;">
+            <div class="card text-center" style="margin-top: 0%;">
```
I do remember mentioning the use of prebuilt Bootstrap margin classes over hardcoded CSS margin statements here https://github.com/fedora-infra/mote/pull/232#pullrequestreview-839715765. Please follow the same and reduce the margin to the `mt-1` scale.

t0xic0der replied:
```suggestion
            <div class="card text-center mt-1">
```

t0xic0der commented on src/mote/templates/mainpage.html:
```diff
@@ -72,12 +79,12 @@
-            <div class="card text-center" style="margin-top: 10%;">
+            <div class="card text-center" style="margin-top: 0%;">
+                <div id="loader"></div>
                 <div class="card-body">
```
```suggestion
                <div class="card-body pt-0">
```

It looks like this now.

![image](https://user-images.githubusercontent.com/49605954/147431831-cdd60afc-3854-4310-a582-598ce92760cd.png)

The padding at the top of the calendar header is not needed. We'll remove it and it would look like the following.

![image](https://user-images.githubusercontent.com/49605954/147431871-7e339fc1-6f03-4483-bbac-01971b95e1a3.png)

t0xic0der reviewed (changes requested):
Here are some more changes that I would want to you to introduce.

t0xic0der reviewed (changes requested):
Please take inspiration from https://calendar.google.com/calendar/u/0/r/week and https://calendar.google.com/calendar/u/0/r/month to make sure that your calendar views occupy upto 100% of the screen height and NOT more.

Having a scrollbar in the calendar views is not only intrusive but it also does not look very elegant in my honest opinion.

t0xic0der reviewed (changes requested):
Please relocate the spinner to be placed right in front of the active "Recent meetings" link in the navbar and then change the link name to "Loading meetings..." until the time the meetings load up. Once they finish loading up, rename it back to "Recent meetings".

subhangi2731 commented:
@t0xic0der sure I am doing the changes

t0xic0der commented:
In the 1:1 that we had on 28th Dec '21, this is a gist of changes that we would want to make happen to the current PR.

![image](https://user-images.githubusercontent.com/49605954/147561645-8d3513d1-769f-4b4e-b7c8-cd6ac89337c3.png)

subhangi2731 commented:
![Screenshot from 2022-01-02 21-04-22](https://user-images.githubusercontent.com/53938490/147880871-cac062e3-ae70-4295-b70a-400d36f0d207.png)

subhangi2731 commented:
@t0xic0der I did all the changes as discussed in the meeting. I made the week view as default and made the option to switch over to the month view for getting recent meetings. I also disabled the arrow buttons for the month view so that it will show the meetings for the current month only and also changed the colors. Also reduced the extra space vertically so that it can come in one scroll. Changing the format to day name and date number one above the other is not possible and is not customizable through any given methods in the docs. Also to disable arrow buttons, I replaced default month and week switching buttons with the custom buttons which changes the view now with the built-in methods and use vanilla JavaScript to disable the arrow buttons when on month view.

subhangi2731 commented:
@t0xic0der please review and let me know.

subhangi2731 commented:
![Screenshot from 2022-01-05 11-43-13](https://user-images.githubusercontent.com/53938490/148169213-3368c609-34b6-4e18-ae0e-268b6ca6852e.png)

subhangi2731 commented:
@t0xic0der @darknao please review my pr and let me know. I have done all the changes as told.

subhangi2731 commented:
![Screenshot from 2022-01-05 13-09-19](https://user-images.githubusercontent.com/53938490/148178898-3aee1864-b731-44e6-84d4-f6cf9e4b5a7e.png)

subhangi2731 commented:
@t0xic0der @darknao I did all the changes please check

subhangi2731 commented:
@t0xic0der @darknao The black lines, I have removed using bootstrap theming so it doesn't raise any incompatibility issues with currently used bootstrap 5. I used custom CSS to style the classes. Grids are now visible in between the days and the calendar adjusts its size according to the screen height

t0xic0der reviewed (approved):
Awesome work. Merging!
